src/mailer/engine.py

```diff
@@ -200,24 +218,17 @@ def send_all():
                         email.connection = None
                         message.email = email  # For the sake of MessageLog
                         MessageLog.objects.log(message, RESULT_SUCCESS)
-                        sent += 1
+                        counts['sent'] += 1
                     else:
                         logging.warning("message discarded due to failure in converting from DB. Added on '%s' with priority '%s'" % (message.when_added, message.priority))  # noqa
                     message.delete()
 
-                except (socket_error, smtplib.SMTPSenderRefused,
-                        smtplib.SMTPRecipientsRefused,
-                        smtplib.SMTPDataError,
-                        smtplib.SMTPAuthenticationError) as err:
-                    message.defer()
-                    logging.info("message deferred due to failure: %s" % err)
-                    MessageLog.objects.log(message, RESULT_FAILURE, log_message=str(err))
-                    deferred += 1
-                    # Get new connection, it case the connection itself has an error.
-                    connection = None
+                except Exception as err:
+                    connection, action = error_handler(connection, message, err)
+                    counts[action] += 1
```

Comment:
PS: I'm still thinking we should have the handler return a bool rather than `sent` and `deferred` — anything but those two values would crash anyway. No?